### PR TITLE
Validation before installing knative serving

### PR DESCRIPTION
**Pull request by yu2003w**
To fix problem #21 

Signed-off-by: Wei Yu(Jared) <yuweiw@cn.ibm.com>

### CODE REVIEW

File: config/crds/serving_v1alpha1_knativeserving_crd.yaml [hunk 1]
```diff
@@ -74,6 +74,17 @@ spec:
                   type: object
                   additionalProperties:
                     type: string
+              knative-dependency:
+                description: Dependencies for installing knative serving, operator need to check whether
```
Comment (mattmoor-sockpuppet) [inline config/crds/serving_v1alpha1_knativeserving_crd.yaml]:
```suggestion
                description: Dependencies for installing knative serving, operator need to check whether
```

Comment (mattmoor-sockpuppet) [inline config/crds/serving_v1alpha1_knativeserving_crd.yaml]:
```suggestion
                description: Dependencies for installing knative serving, operator need to check whether
```

File: config/crds/serving_v1alpha1_knativeserving_crd.yaml [hunk 2]
```diff
@@ -74,6 +74,17 @@ spec:
                   type: object
                   additionalProperties:
                     type: string
+            knative-dependency:
+              description: Dependencies for installing knative serving, operator need to check whether
```
Comment (mattmoor-sockpuppet) [inline config/crds/serving_v1alpha1_knativeserving_crd.yaml]:
```suggestion
              description: Dependencies for installing knative serving, operator need to check whether
```

Comment (mattmoor-sockpuppet) [inline config/crds/serving_v1alpha1_knativeserving_crd.yaml]:
```suggestion
              description: Dependencies for installing knative serving, operator need to check whether
```

**Comment by jcrossley3** [comment]
/hold
I don't think we want dependency versions in the the CR spec. We should already have a mapping of KS release to supported versions of dependencies. Exposing that to the user is just noise, I think.

**Comment by yu2003w** [comment]
@jcrossley3  I agree with you that we should already have the mapping.
But How could serving-operator get the mapping information? Configure that in CM or hard-code for verifying dependency version? 

also CC'ed @trshafer Do you have any advice?

**Comment by trshafer** [comment]
If we're hardcoding the version of knative serving that gets installed, I think it's ok for us to hard code the supported dependencies. It would be nice if these versions were easy to locate within the repo as I imagine people would be interested in looking at what the supported dependency versions were.

**Comment by jcrossley3** [comment]
I would think easy access to dependency version mappings is a question for knative-serving itself to answer, no? The operator would consume that information just like a human would.

**Comment by yu2003w** [comment]
@jcrossley3 Yes, kn serving should know dependencies versions information. I would like to raise a question in kn serving chat group.
Another question is that if I override registry image with kn serving 0.6.1 when creating CR, it could installed successfully. And it seems that we don't have a mechanism to verify that. Is it necessary for serving-operator to validate this?

**Comment by jcrossley3** [comment]
> @jcrossley3 Another question is that if I override registry image with kn serving 0.6.1 when creating CR, it could installed successfully. And it seems that we don't have a mechanism to verify that. Is it necessary for serving-operator to validate this?

Good question. My initial thought is that once you change that image, all bets are off, and you're responsible. It makes me wonder whether we should just disable all dependency checks when an image is overridden.

File: pkg/reconciler/knativeserving/common/validation.go [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+	basemsg := t.meta.name + " validation"
+	// check namespace created
+	_, err := client.CoreV1().Namespaces().Get(t.meta.namespace, metav1.GetOptions{})
+	if err != nil {
+		msg := t.meta.namespace + " doesn't exist"
+		log.Error(err, basemsg, "namespace checking", msg)
+		return fmt.Errorf(msg)
+	}
+
+	// for istio, version could be get frome env of container istio-proxy in pod istio-pilot
```
Comment (mattmoor-sockpuppet) [inline pkg/reconciler/knativeserving/common/validation.go]:
```suggestion
	// for istio, version could be get from env of container istio-proxy in pod istio-pilot
```

Comment (yu2003w) [reply pkg/reconciler/knativeserving/common/validation.go]: > I would think easy access to dependency version mappings is a question for knative-serving itself to answer, no? The operator would consume that information just like a human would.

@jcrossley3 I didn't find information about how to read kn serving mapping with dependency version by operator. Do you have any advice? Or did I misunderstand your comments?

**Comment by yu2003w** [comment]
@jcrossley3 I didn't find information about how to read kn serving mapping with dependency version by operator. Do you have any advice? Or did I misunderstand your comments?

> I would think easy access to dependency version mappings is a question for knative-serving itself to answer, no? The operator would consume that information just like a human would.

File: pkg/reconciler/knativeserving/common/validation.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+	"k8s.io/client-go/tools/remotecommand"
+	"knative.dev/pkg/version"
+	opversion "knative.dev/serving-operator/version"
+)
+
+const (
+	// IstioDependency name
+	istioDependency     = "istio"
+	istioNamespace      = "istio-system"
+	istioProductVersion = "productVersion"
```
Comment (markusthoemmes) [inline pkg/reconciler/knativeserving/common/validation.go]: This seems unused.

File: pkg/reconciler/knativeserving/common/validation.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+	namespace string
+	// deployments installed in the namespace
+	deploy []string
+}
+
+var depMetas = map[string]depMeta{
+	istioDependency: {
+		name:      istioDependency,
+		namespace: istioNamespace,
+		deploy: []string{
```
Comment (markusthoemmes) [inline pkg/reconciler/knativeserving/common/validation.go]: Why do we need this list? we don't really depend on an istio installation even if we depend on istio.

**Review by markusthoemmes — COMMENTED** [review]
I think this should be approached in a more abstract way, at least for Istio. We don't really need to depend on specific deployments being there etc. Really all we depend on at this point is:

1. Some CRDs in a version that we need.
2. The configured gateways (istio-ingressgateway and cluster-local-gateway) installed and at the location defined in Knative's config.

File: pkg/reconciler/knativeserving/common/validation.go (hunk 5 above)
Comment (yu2003w) [reply pkg/reconciler/knativeserving/common/validation.go]: We don't need to know the details of Istio deployments. I will update PR soon.

**Comment by houshengbo** [comment]
Close this PR, because this repo is deprecated.

**Pull request closed without merging**